R/vol-surface-class.R

```diff
@@ -0,0 +1,173 @@
+}
+
+#' @export
+print.VolSurface  <- function(x, ...){
+
+  cat(format(x), "\n")
+  print(x$vol_quotes)
+
+}
+
+
+# VolSurface methods --------------------------------
+
+
+#' Interpolate a `VolSurface` object.
+#'
+#' This method is used to interpolate a `VolSurface` object at multiple points of
+#' the plane. The interpolation depends on the type of the surface, if the vols are
+#' given by strikes, delta, moneyness.
+#'
+#' @param x object of class `VolSurface` to be interpolated.
+#' @param at indicates the coordinates at which the interpolation is performed.
+#' `at` should be given as a named list of length two, where the two members are
+#' vectors with the same length and are named `term` and `strike`. e.g.
+#' list(term = c(1, 2), strike = c(72, 92)).
+#' @param ... unused in this model.
+#' @return `numeric` vector with length equal to the length of `at` members.
+#' @examples vol_surface <- fmbasics::build_vol_surface()
+#' interpolation_points <- list(term = c(as.Date("2020-03-31"), as.Date("2021-03-31")),
+#' strike = c(40, 80))
+#' implied_vols <- fmbasics::interpolate(x = vol_surface, at  = interpolation_points)
+#' @family interpolate functions
+#' @export
+interpolate.VolSurface <- function(x, at, ...) {
+  assertthat::assert_that(
+    all.equal(names(at), c("term", "strike")),
+    length(at$term) == length(at$strike),
+    assertthat::is.date(at$term),
+    is.numeric(at$strike)
+  )
+
+  if (is.LinearTimeVarInterpolation(x$interpolation) & colnames(x$vol_quotes)[2] == "strike") {
+    vols <- x$vol_quotes$value
+    variance <- vols^2
+    tenors <- as.Date(x$vol_quotes$maturity, format = "%Y%m%d")
+    tenors_yf <- fmdates::year_frac(
+      date1 = x$reference_date, date2 = tenors,
+      day_basis = "act/365"
+    )
+
+    time_variance <- variance * tenors_yf
+
+    imp_vol <- rep(NA, length(at$term))
+
+
+    for (i in seq_along(at$term)) {
+      tt <- at$term[i]
+      K <- at$strike[i]
+      smile <- rep(NA, length(unique(x$vol_quotes$strike)))
+      tt_yf <- fmdates::year_frac(
+        date1 = x$reference_date, date2 = tt,
+        day_basis = "act/365"
+      )
+      for (k in 1:length(smile)) {
+        g <- stats::approxfun(
+          x = unique(tenors_yf),
+          y = time_variance[(1 + (k - 1) * length(unique(tenors_yf))):(k * length(unique(tenors_yf)))],
+          method = "linear",
+          rule = 2
+        )
+        smile[k] <- g(tt_yf)
+      }
+      if (K >= min(x$vol_quotes$strike) & K <= max(x$vol_quotes$strike)) {
+        interpolated_time_var <- stats::spline(
+          x = unique(x$vol_quotes$strike),
+          y = smile, method = "natural", xout = K
+        )$y
+      }
+      if (K < min(x$vol_quotes$strike)) {
+        interpolated_time_var <- smile[1]
+      }
+      if (K > max(x$vol_quotes$strike)) {
+        interpolated_time_var <- utils::tail(smile, 1)
+      }
+      imp_vol[i] <- sqrt(interpolated_time_var / tt_yf)
+    }
```

Comment:
I think it may be more efficient (and consistent with `ZeroCurve`) if the calibration interpolator function is stored in the instance of `VolSurface`. Some part of this shouldn't change for a given `(term, strike)` that you are looking to interpolate.

Reply:
This now is the case where the interpolator is embedded in the VolSurface object - calling a separate mathematical Linear_Cubic implemented in utils.R file. Similar to ZeroCurve class logic.